### ISSUE TITLE
Add proper, more self-explanatory errors + type-checking for Block params

### ISSUE THREAD

**Issue by drkameleon** [arturo-lang/arturo#1121]
### Overview

Right now, for standard library functions that take a Block, we perform the usual type-checking (that it's a Block) and that's it.

Now, what if this Block's element have to be of a specific kind? 

Let's take `sum` for example... it takes a Block as a parameter. But could it take a Block of e.g. String values? The answer is obviously *no*. Could it takes a mixed Block? Nope. So in that case, we should trigger an error for the first non-compatible Value found.

In other cases, we may need to perform this type of type-checking even before doing any processing and show the appropriate message beforehand.

And there are other cases, where the block we expect has even stricter characteristics: e.g. `to :complex [...]` takes 2 arguments (not more, not less!) that have to be either Integer's or Floating's. So, that's another one.

**Summing up:**

- we have to create *one*, uniform way of checking for this type of scenarios that will be available throughout the standard library
- we have to - at least try - show all errors, where appropriate

----

**Also see:** https://github.com/arturo-lang/arturo/pull/1117#pullrequestreview-1362557047

**Comment by drkameleon** [comment]
Regarding this scenario:

> Let's take sum for example... it takes a Block as a parameter. But could it take a Block of e.g. String values? The answer is obviously no. Could it takes a mixed Block? Nope. So in that case, we should trigger an error for the first non-compatible Value found.

This is a rather special case, which I'm very tempted to leave as-is right now. (Btw, `sum` & `product` work almost in identical way)

So, how does `sum` work right now?

<img width="343" alt="Screenshot 2023-04-04 at 13 03 26" src="https://user-images.githubusercontent.com/1265028/229772609-c984f02e-62e3-4a44-8d88-0820a379c855.png">

Which may initially seem strange, but is it? Actually `sum` is implemented as a sequential addition of the given block's elements. And addition triggers this type of errors. So, partly it would be normal to expect coming from `sum` directly, but:

- I think the error is understable enough
- Most other interpreted languages would behave in a similar way, showing similar errors (= related to the actual operation being performed; and failing)

Hence, in order not to add extra performance hurdles, I think the wisest choice would be to leave it exactly as it is right now.
